src/ui/public/visualize/loader/pipeline_helpers.js

```diff
@@ -0,0 +1,134 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+
+import { fromExpression } from '@kbn/interpreter/common/lib/ast';
+import { interpretAst } from '@kbn/interpreter/public/interpreter';
+
+const getSchemas = (vis) => {
+  let cnt = 0;
+  const schemas = {
+    metric: [],
+  };
+  const responseAggs = vis.aggs.getResponseAggs();
+  const isHierarchical = vis.isHierarchical();
+  const metrics = responseAggs.filter(agg => agg.type.type === 'metrics');
+  responseAggs.forEach((agg) => {
+    if (!agg.enabled) return;
+    let schemaName = agg.schema ? agg.schema.name || agg.schema : null;
+    if (typeof schemaName === 'object') schemaName = null;
+    if (!schemaName) return;
+    if (!schemas[schemaName]) schemas[schemaName] = [];
+    if (!isHierarchical || agg.type.type !== 'metrics') {
+      schemas[schemaName].push(cnt++);
+    }
+    if (isHierarchical && agg.type.type !== 'metrics') {
+      metrics.forEach(() => {
+        schemas.metric.push(cnt++);
+      });
+    }
+  });
+  return schemas;
+};
+
+const vislibCharts = ['histogram', 'line', 'area', 'gauge', 'goal', 'heatmap', 'horizontal_bar'];
+
+export const buildPipeline = (vis, params) => {
+  const { searchSource } = params;
+  const { indexPattern } = vis;
+  const query = searchSource.getField('query');
+  const filters = searchSource.getField('filter');
+  const visState = vis.getCurrentState();
+
+  let pipeline = `kibana | kibana_context `;
+  if (query) {
+    pipeline += `q='${JSON.stringify(query).replace(/'/g, `\\'`)}' `;
+  }
+  if (filters) {
+    pipeline += `filters='${JSON.stringify(filters).replace(/'/g, `\\'`)}' `;
+  }
+  if (vis.savedSearchId) {
+    pipeline += `savedSearchId='${vis.savedSearchId}' `;
+  }
+  pipeline += '| ';
+  if (vis.type.requestHandler === 'courier') {
```

Comment:
What are we doing in case someone is using a different request handler?

Reply:
custom request handlers/response handlers are not yet supported here. 
we will need a `custom_request_handler` and `custom_response_handler` pipeline functions to which we will fallback. we will also need to come up with the `recommended way` for plugin authors to add this. either they can add their own functions to the pipeline function registry, or we can add some things to the visualization definition.